console_backend/src/ntrip_tab.rs

```diff
@@ -0,0 +1,413 @@
+
+use crossbeam::channel;
+
+use crate::status_bar::Heartbeat;
+
+use anyhow::Context;
+
+use crate::ntrip_output::MessageConverter;
+use crossbeam::channel::Sender;
+use log::error;
+use std::time::{Duration, SystemTime};
+
+#[derive(Debug, Default)]
+pub struct NtripState {
+    pub(crate) connected_thd: Option<JoinHandle<()>>,
+    pub(crate) options: NtripOptions,
+    pub(crate) is_running: ArcBool,
+    last_data: Arc<Mutex<LastData>>,
+}
+
+#[derive(Debug, Default, Copy, Clone)]
+struct LastData {
+    lat: f64,
+    lon: f64,
+    alt: f64,
+    wn: i16,
+    epoch: Option<u64>,
+}
+
+#[derive(Debug, Default, Clone)]
+pub enum PositionMode {
+    #[default]
+    Dynamic,
+    Static {
+        lat: f64,
+        lon: f64,
+        alt: f64,
+    },
+}
+
+#[derive(Debug, Clone, EnumString)]
+pub enum OutputType {
+    RTCM,
+    SBP,
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NtripOptions {
+    pub(crate) url: String,
+    pub(crate) username: Option<String>,
+    pub(crate) password: Option<String>,
+    pub(crate) nmea_period: u64,
+    pub(crate) pos_mode: PositionMode,
+    pub(crate) client_id: String,
+    pub(crate) output_type: Option<OutputType>,
+}
+
+impl NtripOptions {
+    pub fn new(
+        url: String,
+        username: String,
+        password: String,
+        pos_mode: Option<(f64, f64, f64)>,
+        nmea_period: u64,
+        output_type: &str,
+    ) -> Self {
+        let pos_mode = pos_mode
+            .map(|(lat, lon, alt)| PositionMode::Static { lat, lon, alt })
+            .unwrap_or(PositionMode::Dynamic);
+
+        let username = Some(username).filter(|s| !s.is_empty());
+        let password = Some(password).filter(|s| !s.is_empty());
+        NtripOptions {
+            url,
+            username,
+            password,
+            pos_mode,
+            nmea_period,
+            client_id: "00000000-0000-0000-0000-000000000000".to_string(),
+            output_type: OutputType::from_str(output_type).ok(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, serde::Deserialize)]
+#[serde(rename_all = "lowercase")]
+enum Message {
+    Gga { lat: f64, lon: f64, height: f64 },
+}
+
+fn build_gga(opts: &NtripOptions, last_data: &Arc<Mutex<LastData>>) -> Command {
+    let (lat, lon, height, epoch) = match opts.pos_mode {
+        PositionMode::Dynamic => {
+            let guard = last_data.lock().unwrap();
+            (guard.lat, guard.lon, guard.alt, None)
+        }
+        PositionMode::Static { lat, lon, alt } => (lat, lon, alt, None),
+    };
+    Command {
+        epoch,
+        after: 0,
+        crc: None,
+        message: Message::Gga { lat, lon, height },
+    }
+}
+
+#[derive(Debug, Clone, Copy, serde::Deserialize)]
+struct Command {
+    #[serde(default = "default_after")]
+    after: u64,
+    epoch: Option<u64>,
+    crc: Option<u8>,
+    #[serde(flatten)]
+    message: Message,
+}
+
+fn default_after() -> u64 {
+    10
+}
```

Comment:
better to define this as a const val, or at least `const fn` if you really want to use fn